feat(server): add support for multiple auth tokens via env vars

Nydragon opened the pull request:
<!--- Thank you for contributing to rustypaste! -->

## Description

Add support for reading a list of newline separated tokens from a file
which path is being supplied through the following environment
variables:

- AUTH_FILE
- DELETE_FILE

Also add support for setting environment variables inside of fixtures through the optional `custom_env` function.

Also describe this new functionality in the readme.

<!--- Describe your changes in detail -->

## Motivation and Context

Need a way to supply multiple auth token akin to `AUTH_TOKEN` due to agenix behaviour.

closes #338

<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->

## How Has This Been Tested?

<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

Added UTs for file content parsing. File reading has not received any UT, but has been manually tested.

## Changelog Entry

<!--- Please write the changelog entry for these changes. -->
<!--- Follow the <https://keepachangelog.com/en/1.0.0/> format. -->
<!--- Use one of the Added, Changed, Deprecated, Removed, Fixed, and Security headers accordingly. -->
<!-- For example:
````
### Added

- Add a middleware for checking the content length
  - Before, the upload size was checked after full upload which was clearly wrong.
````
-->

### Added

- Multiple authentication and deletion token parsing via a file to which `AUTH_FILE` and `DELETION_FILE` point. This file needs to be a newline separated list of strings. Whitespace is discarded.

## Types of Changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] Documentation (no code change)
- [ ] Refactor (refactoring production code)
- [ ] Other <!--- (provide information) -->

## Checklist:

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->

- [x] My code follows the code style of this project.
- [x] I have updated the documentation accordingly.
- [x] I have formatted the code with [rustfmt](https://github.com/rust-lang/rustfmt).
- [x] I checked the lints with [clippy](https://github.com/rust-lang/rust-clippy).
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

tessus reviewed (approved):
Thanks for this PR. LGTM (unless I missed some Rust specific optimization I am not aware of, since I am not a Rust expert).

Nydragon commented:
> Oh, could you add a test to the fixtures as well?

Sure! I didn't see an option to set env vars tho... Did i miss something?

tessus commented:
> I was wondering if it would be better to `.trim()` the entire file or the individual lines after splitting (current behaviour).

One would have to run perf tests for the difference, but I doubt it is huge for a small to medium sized file with tokens.
Personally I like the current behavior better, since you took care of a few edge cases as well.

The only workflow optimization I can think of is that the file is not read and parsed every time a request is made, but only when the config is read or reloaded via `refresh_rate`.

If you want to try that, please go ahead.

Nydragon commented:
> > I was wondering if it would be better to `.trim()` the entire file or the individual lines after splitting (current behaviour).
> 
> One would have to run perf tests for the difference, but I doubt it is huge for a small to medium sized file with tokens. Personally I like the current behavior better, since you took care of a few edge cases as well.
> 
> The only workflow optimization I can think of is that the file is not read and parsed every time a request is made, but only when the config is read or reloaded via `refresh_rate`.
> 
> If you want to try that, please go ahead.

That'd make definitely sense! I was a bit surprised to see that `get_tokens` is called on every request. Why is that? Wouldn't it be enough to just call the function periodically according to `refresh_rate`?

tessus commented:
> That'd make definitely sense! I was a bit surprised to see that `get_tokens` is called on every request. Why is that? Wouldn't it be enough to just call the function periodically according to `refresh_rate`?

This is a good question. I think it developed this way. At the beginning there was only one token in the config file and one env var. I added the multiple tokens and delete tokens, and then it was refactored by somebody else to be used as a middleware with web grants.
In all these stages only the config was referenced (not the file read) and the ENV vars read, thus no real overhead.

orhun commented on src/config.rs:
```diff
@@ -170,6 +180,16 @@ impl Config {
+
+                if let Ok(env_path) = env::var(DELETE_FILE_ENV) {
+                    match read_to_string(&env_path) {
+                        Ok(s) => push_token_file(s, &mut tokens),
```
No need for a function call:

```suggestion
                        Ok(s) => {
                            s.lines().filter(|l| !l.trim().is_empty()).for_each(|l| {
                                tokens.insert(l.to_string());
                            });
                        }
```

Nydragon replied:
I originally created a function to be able to write a UT for the parsing, is just a fixture fine then ?

orhun replied:
Yup, it should be enough.

orhun commented on src/config.rs:
```diff
@@ -170,6 +180,16 @@ impl Config {
+                    match read_to_string(&env_path) {
+                        Ok(s) => push_token_file(s, &mut tokens),
+                        Err(e) => {
+                            error!("Failed reading from deletion file {env_path}.\n{e}");
```
```suggestion
                            error!("failed to read deletion tokens from file ({env_path}) ({e})");
```

orhun commented on src/config.rs:
```diff
@@ -162,6 +163,15 @@ impl Config {
                 }
+                if let Ok(env_path) = env::var(AUTH_FILE_ENV) {
+                    match read_to_string(&env_path) {
+                        Ok(s) => push_token_file(s, &mut tokens),
```
No need for a function call.

```suggestion
                        Ok(s) => {
                            s.lines().filter(|l| !l.trim().is_empty()).for_each(|l| {
                                tokens.insert(l.to_string());
                            });
                        }
```

orhun commented on src/config.rs:
```diff
@@ -203,6 +223,18 @@ impl Config {
     }
 }
 
+fn push_token_file(string: String, tokens: &mut HashSet<String>) {
+    string
+        .split('\n')
+        .filter_map(|s| match s.trim() {
+            "" => None,
+            s => Some(s),
+        })
+        .for_each(|t| {
+            tokens.insert(t.to_string());
+        });
+}
+
```
Not needed.

orhun reviewed (commented):
Looks good! We can simplify it a bit :)

orhun commented:
> are you ok with adding another hook to add env vars to the fixtures?

Hmm, yeah, if it is not much of a hassle. Bash is annoying.

>  I was a bit surprised to see that get_tokens is called on every request. Why is that? 

We use a library called `actix_web_grants` which works by adding a middleware and then you can simply protect your endpoints with an annotation.

https://github.com/orhun/rustypaste/blob/1d5a9c60044ab0ece53868eccbe7cf011d0c5633/src/server.rs#L417

https://github.com/orhun/rustypaste/blob/1d5a9c60044ab0ece53868eccbe7cf011d0c5633/src/server.rs#L153

Implemented in #199 

That is why `extract_tokens` is called on every request since there is a guard. We can maybe cache some parts to make it more performant.

> Wouldn't it be enough to just call the function periodically according to refresh_rate?

That means we need to store the tokens somewhere (probably in memory) and that's not how the grants library works.

Nydragon commented:
Alright, thank you for the explanation, should I then leave out the improvements mentioned by tessus (https://github.com/orhun/rustypaste/pull/339#issuecomment-2304652210)?

I'll also add the env handling to the fixture script.

Nydragon commented:
Should be all good now:

- added fixture for the file based authentication process
- added fixture for the file based deletion process
- added env var setting through the `rp_env` function, teardown is handled by putting `run_fixture` into a subshell
- applied code review requests

orhun commented on fixtures/README.md:
```diff
@@ -28,6 +28,11 @@ test-file-upload/
 #!/usr/bin/env bash
 
+# Optional
+rp_env() {
```
I think it would be better if we rename this to `custom_env`.

orhun commented on fixtures/README.md:
```diff
@@ -28,6 +28,11 @@ test-file-upload/
 
+# Optional
+rp_env() {
+  # setting env variables such as AUTH_TOKEN or AUTH_FILE
```
nit:
```suggestion
  # setting environment variables such as AUTH_TOKEN or AUTH_FILE
```

orhun commented on fixtures/test-fixtures.sh:
```diff
@@ -9,11 +9,14 @@ NC='\033[0m'
 run_fixture() {
   cd "$FIXTURE_DIR/$1" || exit 1
   source "test.sh"
+  # Check if rp_env exists
```
it's clear, no need for a comment here :)

orhun commented on fixtures/test-fixtures.sh:
```diff
@@ -24,7 +27,9 @@ run_fixture() {
 main() {
   find * -maxdepth 0 -type d -print0 | while IFS= read -r -d '' fixture; do
-    run_fixture "$fixture"
+    # Since we are creating a subshell, all env vars created by rp_env will be lost
+    # Return code is preserved
+    (run_fixture "$fixture")
```
Is there any side effects of this? Are we sure that the return code is preserved?

I'd test it manually with a failing fixture on purpose to see if the behavior is the same.

also nit:

```suggestion
    # Since we are creating a subshell, all environment variables created by rp_env will be lost
    # Return code is preserved
    (run_fixture "$fixture")
```

tessus replied:
Interesting. I would have used `custom_env_teardown()` and added it to `run_fixture()` before retuning.

But if the subshell works, it's certainly a better option.

Nydragon replied:
I echo'd the return value during my modifications and it did preserve the return code properly. As for any side effects I couldn't notice any.

<details><summary>Fixtures run normally:</summary>
<p>

![image](https://github.com/user-attachments/assets/b3796f7a-d794-45c2-92c0-b9668d0c6cfb)
</p>
</details> 

<details><summary>Modified `test` to fail</summary>
<p>

![image](https://github.com/user-attachments/assets/57346884-5545-4e35-8b73-d7bda4553439)

</p>
</details>

Nydragon commented:
Applied all requests

tessus reviewed (commented):
Great, thanks.

One last thing. Could you please update the README in the fixtures dir?

Can you please add the new function to the file format?

Nydragon commented:
I think I already did that: https://github.com/orhun/rustypaste/pull/339/files#diff-3374186a79a0a6316047cb823494f238bdc644489b1b709bf65c192b5c05183a

tessus commented:
Thanks, I missed it. All good then! I think it is ready to be merged.

orhun commented on src/lib.rs:
```diff
@@ -41,5 +41,11 @@ pub const CONFIG_ENV: &str = "CONFIG";
 pub const DELETE_TOKEN_ENV: &str = "DELETE_TOKEN";
+
+/// Environment variable for the path to a file containing multiple deletion token.
+pub const DELETE_FILE_ENV: &str = "DELETE_FILE";
```
I was about to merge this but then had an epiphany, `DELETE_FILE` sounds more like "delete this file" rather than "a file that contains the deletion tokens". I think we should grammatically fix this while we have the chance :) So I propose:

- `AUTH_FILE` -> `AUTH_TOKENS_FILE`
- `DELETE_FILE` -> `DELETE_TOKENS_FILE`

Thoughts? @Nydragon @tessus

Nydragon replied:
Completely agree, actually chose that name at the beginning, but replaced it with the current one because it was shorter.

Will update that as soon as I'm back home!

tessus replied:
Shorter names only make sense, if it has to be used all the time....

Nydragon replied:
Changed the names

orhun reviewed (approved):
Perfect, thanks!
